runtime/lua/vim/treesitter/language.lua

```diff
@@ -23,15 +23,21 @@ function M.require_language(lang, path, silent, symbol_name)
       error("'" .. lang .. "' is not a valid language name")
     end
 
-    local fname = 'parser/' .. lang .. '.*'
-    local paths = a.nvim_get_runtime_file(fname, false)
-    if #paths == 0 then
+    for _, l in ipairs {lang, symbol_name} do
```

Comment:
I'd rather have the user decide what they want instead of nvim trying out multiple paths. If the SO-filename should be configurable, it could be an extra parameter.

As I understood https://github.com/nvim-treesitter/nvim-treesitter/pull/4296#issuecomment-1426995523, this to enable `ft` to `lang` mapping like the Lua table that we have in nvim-treesitter. Please note that when you're moving the functionality of mapping `ft` to parser to the `treesitter.c` hashmap, you will load the same parser multiple times as the same SO-parser file can be loaded for multiple `lang`s (`lang` being the key for a loaded parser). Isn't this something that the caller of `require_language` can handle via a simple Lua table mapping? At nvim-treesitter, we always tried to have `ft` and `lang` be separate concepts.

Reply:
> I'd rather have the user decide what they want instead of nvim trying out multiple paths. If the SO-filename should be configurable, it could be an extra parameter.

It already is as the second `path` argument.

> Please note that when you're moving the functionality of mapping ft to parser to the treesitter.c hashmap, you will load the same parser multiple times as the same SO-parser file can be loaded for multiple langs (lang being the key for a loaded parser).

I'm aware of this, and this is just a small implementation deficiency that should be easy enough to resolve. In practice, this will just mean that users with a `bash` and `zsh` buffer open in the same session will have the `sh` parser loaded multiple times. For the majority, this shouldn't be an issue since `filetype` usually maps directly to `lang`.

> At nvim-treesitter, we always tried to have ft and lang be separate concepts.

I've seen no benefit in keeping this distinction, and for me has only ever caused problems.

For example, I've currently had to introduce some awkward logic in https://github.com/neovim/neovim/pull/22191 because `get_parser` defaults to `filetype` for `lang` despite a parser already being associated for a buffer.

The end goal I'm working to is that the whole `filetype` -> `lang` relationship is encapsulated to parser installation and management. Everywhere else in any treesitter logic, only filetype should be considered.

In this PR I've only looked in both paths (for when `path` isn't provided) just to maintain some compatibility for users who will continue to pass `lang` as the first argument. Once downstream has adapted, which we won't give much time for, we then only search for parsers in `parser/{symbol_name}.so`

More context, I believe we were forced to maintain the `filetype` - `lang` distinction because of initial design mistakes in nvim's `treesitter.c` which was recently resolved with https://github.com/neovim/neovim/pull/19496.